docs/design_notes/module_lifecycle.md

```diff
@@ -0,0 +1,289 @@
+# The Module Lifecycle
+
+This design note is intended for people who are somewhat familiar with Modules but want to understand more about the design principles behind the abstraction. After reading this, you should have a solid understanding of the assumptions and guarantees the Module API is built upon. If you have no practical experience with Modules yet, Please take a look at the [MNIST tutorial](https://flax.readthedocs.io/en/latest/notebooks/annotated_mnist.html).
```

Comment:
More suggestions:

```suggestion
This design note is intended for users who are already familiar with Flax Modules and want to understand the design principles behind the abstraction. This note should give you a strong understanding of the assumptions and guarantees the Module API is built upon. If you have no practical experience with Modules, check out the [MNIST tutorial](https://flax.readthedocs.io/en/latest/notebooks/annotated_mnist.html) to get started.
```

Note: https://developers.google.com/style/translation?hl=en#be-inclusive (writing for a global audience)

Reply:
I think it applies for example to the line below: "off the beaten track". The link is interesting but also makes things more complicated :) One thing I observed is that many communities have moved to a much more engaged/personal/funny way of writing docs rather than the dry reference material that was common before. Yet what is funny for one person is confusing to another. And what is boring for some people is clear and to the point for others.

Reply:
@marcvanzee I think it's things like "colloquialisms, idioms, or slang" or things that may come across as such to users who aren't as proficient in English/"computer science" English (https://developers.google.com/style/translation)

@jheek Yup, it's not easy to balance